Add additional events and better event args

Ruhrpottpatriot opened the pull request:
This pull request added additional events for errors, open and close. These events and the OnMessage event now use the `EventArgs` or `EventArgs<T>` type to pass along information to the subscriber.
This allows the user to manage multiple socket connections and differentiate by sender.

I also changed all event handler invocations (except `Close`) to be asynchronous, which was done with the `Task.Factory.FromAsync()` method.

Last I added a method to allow the user to add his own headers to the initial socket request. This allows greater flexibility with things such as authorization,

Ruhrpottpatriot committed:
Add additional event handlers

Adds event handlers for Open, Close and Error. Error also passes
the exception in its event args.
The event handler for message received was rewritten to use proper
event args.
Furthermore all event handlers are now invoked asynchronously.

Ruhrpottpatriot committed:
Update README

Ruhrpottpatriot committed:
Add option to add custom request headers

Ruhrpottpatriot committed:
Rename event handlers

Ruhrpottpatriot committed:
Improve documentation

Ruhrpottpatriot committed:
Merge remote-tracking branch 'upstream/master'

Ruhrpottpatriot committed:
Add constructur to make CancellationToken optional

When making the CancellationToken parameter optional in the constructor
it needs to be boxed. A second constructor with constructor chaining,
passing an empty token will avoid boxing and unboxing, thus avoiding
the performance penalty

Ruhrpottpatriot committed:
Update Readme

dimaaan commented on WebSocketTextClient/WebSocketTextClient.cs:
```diff
@@ -1,119 +1,163 @@
+        /// <summary>Signals that response message fully received and ready to process.</summary>
+        public event EventHandler<MessageReceivedEventArgs> MessageReceived;
+
+        /// <summary>Singals that the websocket received an error.</summary>
```
Typo! :)

Ruhrpottpatriot replied:
Whoops! Lemme fix that.

Ruhrpottpatriot committed:
Revert asynchronous event handler invocation

Asynchronous handler invocation throws an InvalidOperationException.
Thus handlers are now invoked sychronously (again).

Ruhrpottpatriot commented:
Btw: I intended this as a starting point. There's still more room for improvement. For example, I find it a bit counter intuitive, how you handled the keep alive process. If I understood it correctly, you call `OpenAnsyc()` and that Task only completes when the connection is closed.

Maybe change the code, so that every public async method accepts a `CancellationToken` and the receive Task is decoupled from `ConnectAsync()`.

`TaskCompletionSource` could be an option. I'll look into it.

Ruhrpottpatriot committed:
Fixed some typos

dimaaan commented:
> I intended this as a starting point. There's still more room for improvement...

Currently, cancel a task via token is the only way to close connection.
Agree that this should be improved.
I'll create a separate issue for this

dimaaan commented on WebSocketTextClient/WebSocketTextClient.cs:
```diff
@@ -123,7 +123,7 @@ public Task SendAsync(string str)
             catch (Exception ex)
             {
-                this.ErrorReceived?.Invoke(this, new SocketErrorEventArgs { Exception = ex, Message = string.Empty });
+                this.ErrorReceived?.Invoke(this, new SocketErrorEventArgs { Exception = ex, Message = "An error occoured." });
```
Yet another typo

dimaaan replied:
Actually, i don't see any value in SocketErrorEventArgs.Message field. We already have error description inside Exception property. Can you remove it?

Ruhrpottpatriot replied:
You usually set the message property to a human readable message, since sometimes you are not getting the correct exception, although afaik, this is usually encountered in WPF.

I can remove it, sure.

dimaaan replied:
Yes, remove it please.
I guess error message for end user should be set in hosting application, not in the library. This way we can handle localization, etc.
Not sure about WPF, i use it in console app/windows service

Ruhrpottpatriot replied:
I want to add, that the logging library of netcore (Microsoft.Extensions.Logging) the log warning and log critical methods have the following header: `LogWarning(Exception exception, string message, params string[] args);` and `LogCritical(Exception exception, string message, params string[] args);`

Ruhrpottpatriot committed:
Add close support

Currently the only way to "close" the socket was a CancellationToken
passed in the contructor. However, this did only abort the running tasks
and did not properly close the uncerlying connection.
Additionally a call to Dispose() is not guaranteed and should not be
responsible for connection closing.
The DisconnectAsync() closes the connection regardless of GC and aborts
all currently running tasks.

Ruhrpottpatriot committed:
Remove 'Message' property from SocketErrorEventArgs

dimaaan reviewed (approved):
Thanks for contribution!
